Hosting a Directus Application on an Ubuntu Server

koredeycode opened the pull request:
Hi.

I'm submitting my article for your review.

Open to feedback.

Thank you.

Closes #14

phazonoverload commented on ubuntu-deploy/index.md:
```diff
@@ -0,0 +1,414 @@
+---
+title: 'Hosting a Directus Application on an Ubuntu Server'
```
```suggestion
title: 'Deploy Directus on an Ubuntu Server with Docker'
```

Just to make it more consistent with other posts

phazonoverload commented on ubuntu-deploy/index.md:
```diff
@@ -0,0 +1,414 @@
+
+## Introduction
+
+In this tutorial, you will learn how to deploy a Directus application within a Docker container on an Ubuntu Linux server and connect it to a custom domain, such as `https://directus.exampledomain.com`.
```
```suggestion
In this tutorial, you will learn how to deploy a Directus application within a Docker container on an Ubuntu server and connect it to a custom domain. Ubuntu is a popular open source Linux distribution  which is commonly available from hosting providers. 
```

Adding a one-liner about what Ubuntu is just in case folks don't know.

phazonoverload commented on ubuntu-deploy/index.md:
```diff
@@ -0,0 +1,414 @@
+
+2. **Ubuntu Server (Version 20.04 or 22.04):** Access an Ubuntu server via SSH. Obtain one from cloud providers like Azure, DigitalOcean, Linode, or AWS. Configure SSH access from your local machine.
+
+3. **Domain Name:** Register a domain name and have access to its DNS settings. This is crucial for the domain to resolve to your Ubuntu server
```
```suggestion
3. **Domain Name:** Register a custom domain name and have access to its DNS settings. This is crucial for the domain to resolve to your Ubuntu server.
```

phazonoverload commented on ubuntu-deploy/index.md:
```diff
@@ -0,0 +1,414 @@
+
+3. **Domain Name:** Register a domain name and have access to its DNS settings. This is crucial for the domain to resolve to your Ubuntu server
+
+4. **Basic Command-Line Skills:** Basic familiarity with Linux command-line operations, including file uploads using `scp` and editing files with `vi`.
```
```suggestion
4. **Command Line Familiarity:** Basic knowledge of Linux command-line operations, including file uploads using `scp` and editing files with `nano`.
```

Let's keep this post easy and use the other built-in text editor `nano`.

phazonoverload commented on ubuntu-deploy/index.md:
````diff
@@ -0,0 +1,414 @@
+
+Assuming you have a local Directus folder structure that looks like this:
+
+```plaintext
+├── database
+│   └── data.db
+├── docker-compose.yml
+├── uploads
+└── extensions
+```
+
+And the _docker-compose.yml_ file looks somewhat like this:
+
+```yaml
+version: '3'
+services:
+  directus:
+    image: directus/directus:latest
+    ports:
+      - 8055:8055
+    volumes:
+      - ./database:/directus/database
+      - ./uploads:/directus/uploads
+      - ./extensions:/directus/extensions
+    environment:
+      KEY: 'replace-with-random-value'
+      SECRET: 'replace-with-random-value'
+      ADMIN_EMAIL: 'admin@example.com'
+      ADMIN_PASSWORD: 'd1r3ctu5'
+      DB_CLIENT: 'sqlite3'
+      DB_FILENAME: '/directus/database/data.db'
+      WEBSOCKETS_ENABLED: 1
+      # Add your other settings
+```
+
+:::info Security Note
+
+Ensure you adjust the email and set a strong password for the admin user.
+
+:::
````
Because this is exactly the same as the quickstart guide, and we don't later edit the `docker-compose.yml` file, we can link our and summarize here. Something like: 

> If you have successfully followed the [Self-Hosted Quickstart][quickstart], you should have a directory with a `docker-compose.yml` file, `database/` directory with a `data.db` file, `uploads/` directory, and `extensions/` directory.

phazonoverload commented on ubuntu-deploy/index.md:
```diff
@@ -0,0 +1,414 @@
+
+In the example below, I am copying the whole Directus folder to the home directory(`~`) of my server.
+
+![Copying files to the server][image-6]
```
Can you briefly explain why only the data.db and compose file are shown in the terminal here?

koredeycode replied:
I think it is to improve readability and avoid cluttering the output. One can easily identify the files being transferred without displaying the entire directories

phazonoverload commented on ubuntu-deploy/index.md:
````diff
@@ -0,0 +1,414 @@
+
+This command updates and upgrades all existing packages to their latest versions.
+
+## Installing Docker and Docker Compose
+
+To install Docker, run the following commands:
+
+```bash
+sudo apt install docker.io -y
+sudo systemctl start docker
+sudo systemctl enable docker
+```
+
+This installs Docker and enables the Docker service.
+To check if Docker is running, run:
+
+```bash
+sudo systemctl status docker
+```
+
+To install Docker Compose, run the following commands:
+
+```bash
+sudo apt install docker-compose -y
+```
+
+To verify that the installation was successful, you can run:
+
+```bash
+docker-compose --version
+```
````
This is out of date according to this https://docs.docker.com/engine/install/ubuntu/ `docker.io` is an unofficial package. Can we update the instructions incl the installation of docker compose?

phazonoverload commented on ubuntu-deploy/index.md:
````diff
@@ -0,0 +1,414 @@
+sudo docker-compose up
+```
+
+![Running docker compose][image-11]
````
I don't think this adds much value and would encourage you to remove it.

phazonoverload commented on ubuntu-deploy/index.md:
```diff
@@ -0,0 +1,414 @@
+
+Your application should now be accessible at `http://your_server_ip:8055`.
+
+![Directus application accessed with the server ip at port 8055][image-2]
```
I don't think this adds much value and would encourage you to remove it.

phazonoverload commented on ubuntu-deploy/index.md:
````diff
@@ -0,0 +1,414 @@
+
+![Directus application accessed with the server ip at port 8055][image-2]
+
+:::info Error Debugging
+
+If you encounter any error e.g `SQLITE_CANTOPEN: unable to open database file`, it is probably due to permission issues. You can try running this command to fix it:
+
+```bash
+sudo chown -R $USER:$USER /path/to/your/directus/folder
+
+```
+
+More info about this error in this [issue][issue]
+
+:::
````
````suggestion
:::info Debugging SQLITE_CANTOPEN

If you encounter any error e.g `SQLITE_CANTOPEN: unable to open database file`, it is probably due to [permission issues][issue]. You can run this command to fix it:

```bash
sudo chown -R $USER:$USER /path/to/your/directus/folder
```

:::
````

A bit more concise

phazonoverload commented on ubuntu-deploy/index.md:
```diff
@@ -0,0 +1,414 @@
+
+## Running the Docker Container as a Background Service
+
+Running your application like above is fine, but it will stop running when you close the terminal.
```
```suggestion
Running your application using `sudo docker-compose up` will stop it running when you close the terminal.
```

phazonoverload commented on ubuntu-deploy/index.md:
````diff
@@ -0,0 +1,414 @@
+Create a file named _directus.service_ in the _/etc/systemd/system/_ directory:
+
+```bash
+sudo vi /etc/systemd/system/directus.service
````
```suggestion
sudo nano /etc/systemd/system/directus.service
```

phazonoverload commented on ubuntu-deploy/index.md:
````diff
@@ -0,0 +1,414 @@
+sudo vi /etc/systemd/system/directus.service
+```
+
+Add the following content, updating the `WorkingDirectory` to your Directus project directory containing the _docker-compose.yml_ file:
````
```suggestion
Add the following content, updating the `WorkingDirectory` to your Directus project directory containing the `docker-compose.yml` file:
```

phazonoverload commented on ubuntu-deploy/index.md:
```diff
@@ -0,0 +1,414 @@
+
+You can get the full path to you directory by running the command `pwd` in the project directory on your server and copying the output.
+
+![print working directory(pwd) command][image-10]
```
I don't think this image adds much value and would encourage you to remove it. Your description and the visual callout is more than enough.

phazonoverload commented on ubuntu-deploy/index.md:
```diff
@@ -0,0 +1,414 @@
+
+Let's step through the service file:
+
+- [Unit] Section:
+
+  - Description: Description of the systemd service.
+  - Requires: Dependency on the Docker service.
+  - After: Starts after the Docker service.
+
+- [Service] Section:
+
+  - Restart=always: The service restarts automatically on exit.
+  - WorkingDirectory: Path to the directory containing the docker-compose.yml file.
+  - ExecStart: Command to start Docker containers.
+  - ExecStop: Command to stop Docker containers.
+
+- [Install] Section:
+  - WantedBy=multi-user.target: Service enabled on reaching the multi-user state.
```
Great summary! As someone who hasn't used systemd before this was very clear.

phazonoverload commented on ubuntu-deploy/index.md:
````diff
@@ -0,0 +1,414 @@
+sudo systemctl start directus.service
+```
+
+Now, your dockerized Directus application is running as a background service and will automatically restart in case of failures or system reboots.
````
I'd love you to add an explicit note if this works across system restarts as well

phazonoverload commented on ubuntu-deploy/index.md:
```diff
@@ -0,0 +1,414 @@
+## Configuring DNS Settings for Your Domain
+
+Configuring DNS settings for your domain is a crucial step in making your Directus application accessible to users over the internet. Here's how to do it:
+
+1. Access Your Domain Registrar's Website: Log in to the website of your domain registrar, where you initially purchased or registered your domain name. This is where you manage your domain settings.
+
+2. Locate DNS Management or Domain Settings: Inside your domain registrar's dashboard, look for options like "DNS Management," "Domain Settings," or "Domain Management." These names might vary based on the registrar's interface.
+
+![DNS management dashboard][image-8]
+
+3. Add a DNS Record for Your Subdomain: Create a new DNS record to point your subdomain (e.g., directus.exampledomain.com) to your server's public IP address. Depending on the registrar, you may need to choose the record type, which is usually an "A" record for IPv4 addresses. Enter your server's public IP address in the designated field.
+
+![A record domain configuration][image-5]
+
+4. Save the changes: After adding the DNS record, save the changes. DNS propagation might take some time, ranging from a few minutes to a few hours. During this period, the DNS changes will propagate across the internet, making your subdomain accessible.
+
+You can confirm your changes by visiting the application by visiting `http://directus.exampledomain.com:8055`.
```
This is good. Let's remove the images as they are VERY vendor-specific. But the description makes sense.

phazonoverload commented on ubuntu-deploy/index.md:
```diff
@@ -0,0 +1,414 @@
+
+## Setting Up Nginx as a Reverse Proxy
+
+Nginx, often called 'engine-x,' is a powerful reverse proxy server widely used in web hosting.
+
+As a reverse proxy, Nginx sits between clients and backend servers, forwarding client requests to the appropriate server.
```
Can we add another line or two explaining the 'why' of Nginx? I also think we can remove the pronunciation guide as it's unimportant and we haven't done it elsewhere.

phazonoverload commented on ubuntu-deploy/index.md:
````diff
@@ -0,0 +1,414 @@
+Create an Nginx configuration file named _directus_ for your domain:
+
+```bash
+sudo vi /etc/nginx/sites-available/directus
````
```suggestion
sudo nano /etc/nginx/sites-available/directus
```

phazonoverload commented on ubuntu-deploy/index.md:
````diff
@@ -0,0 +1,414 @@
+}
+```
+
+This simply forward the request to the domain to the Directus application running at `localhost` port `8055`.
````
```suggestion
This forwards the request to the domain to the Directus application running at `localhost` port `8055`.
```

Removing 'simply' for [this reason](https://justsimply.dev/).

phazonoverload commented on ubuntu-deploy/index.md:
````diff
@@ -0,0 +1,414 @@
+sudo nginx -t
+```
+
+Create a symbolic link to enable the site:
````
Again a quick explainer of what's happening here would be good.

phazonoverload commented on ubuntu-deploy/index.md:
```diff
@@ -0,0 +1,414 @@
+
+Now you should be able to access your Directus application without adding the port at `http://directus.exampledomain.com`.
+
+![Directus application accessed at the domain but insecure][image-1]
```
I don't think this image adds much value and would encourage you to remove it.

phazonoverload commented on ubuntu-deploy/index.md:
```diff
@@ -0,0 +1,414 @@
+
+## Securing Your Application with SSL (Optional but Recommended)
+
+Implementing SSL (Secure Sockets Layer) encryption is crucial for safeguarding data transmitted between your users and the server. [Let's Encrypt][let-encrypt] offers free SSL certificates, and here's how to set it up for your Directus application:
```
```suggestion
Implementing SSL (Secure Sockets Layer) encryption is crucial for safeguarding data transmitted between your users and the server. Once enabled, your application will be accessible using `https`. [Let's Encrypt][let-encrypt] offers free SSL certificates, and here's how to set it up for your Directus application:
```

Explaining the end-state upfront.

phazonoverload commented on ubuntu-deploy/index.md:
```diff
@@ -0,0 +1,414 @@
+
+![HTTP to HTTPS redirect with certbot][image-9]
+
+:::info Renew certificate
+
+Also, ensure to renew the certificate before it expires to maintain a secure connection.
+
+:::
```
Why an info box? I'd just add it as text here.

phazonoverload commented on ubuntu-deploy/index.md:
```diff
@@ -0,0 +1,414 @@
+
+3. Verify SSL Configuration: After the setup is complete, visit your Directus application using `https://directus.exampledomain.com` in a web browser. You should see a padlock icon indicating a secure SSL connection. You should also be automatically redirected from `http` to `https`.
+
+![Secured Directus application][image-3]
```
I'll crop this when I upload it to remove whitespace below the popup, but really useful.

phazonoverload reviewed (changes requested):
Excellent post - a really good read. I've added quite a few comments of things to adjust - there are a few areas that need a tiny bit more explanation, and a few images I think can be removed.

koredeycode commented:
Thanks for the feedbacks @phazonoverload. 

I have made the changes as request, updated the installation instruction for docker and remove and updated the images.

phazonoverload commented:
Thank you for writing this, and I'm happy to consider it done in it's current state. I'll schedule it in to be published either this Thursday or next Tuesday.

This repo is just for authoring/reviewing - the actual piece will be published via a Directus project, so I'll take it from here. I'll close this issue to keep the main branch clean, but this is a acceptance of your work.

Could you email me at [kevin@directus.io](mailto:kevin@directus.io) and I'll share what we need to get you paid.

I'm very excited to publish this piece - it's very cool.
